### PR TITLE
Calculator -- Dessert version

### PR DESCRIPTION
**Pull request by ktrops**
Here is the dessert version of my calculator.

### CODE REVIEW

**Comment by khjrtbrg** [comment]
Starting code review!

File: calc.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    next
+  end
+  #Changes answer to a float if necessary. 
+  if answer.to_f % 1 == 0
+    
+    num1 = num1.to_i
+    num2 = num2.to_i
+    answer = answer.to_i
+  end
+    puts "The result is #{num1.to_s + " " + operation + " " + num2.to_s + " = " + answer.to_s} "
```
Comment (khjrtbrg) [inline calc.rb]: :+1: Nice! I like how you used variables to output a string at the end rather than generating a string inside each conditional.

You're already using interpolation on this line, so I'd keep using it instead of concatenating within the interpolated section. Since the final output will be a string, you also don't need to convert your integer variables to strings beforehand when interpolating in Ruby.

``` ruby
puts "The result is #{num1} #{operation} #{num2} = #{answer}"
```

It also looks like you have an extra indent before this line that you can remove.

File: calc.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
+answer = nil
+puts "Hi there"
+while answer == nil 
+  puts "Please enter one of the following operations, addition, subtraction, \nmultiplication, division, exponent, or modulo:"
+  operation = gets.chomp.downcase
+#gets user input
+  puts "Please enter the first number:"
+  user_num1 = gets.chomp
+  num1 = user_num1.to_f
```
Comment (khjrtbrg) [inline calc.rb]: It looks like you end up converting this and user_num2 to integers to do some checks in the next if statement and at the very end. Is there a reason why you chose to use floats first and then convert it to integers later? Your code will be more consistent if you stick with one or the other.

File: calc.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  
+  puts "Please enter the second number:"
+  user_num2 = gets.chomp
+  num2 = user_num2.to_f
+#test to see if user input is acutally a number
+  if user_num1.to_i == 0 && user_num1 != "0"
```
Comment (khjrtbrg) [inline calc.rb]: You could have used `num2` here to avoid converting `user_num1` inline and take advantage of variables you already created.

File: calc.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    puts user_num1.to_i
+    puts "You did not enter a number. You entered #{user_num1.class} or #{user_num2.class} type."
+    next
+  elsif user_num2.to_i == 0 && user_num2 != "0"
+    puts "You did not enter a number. You entered #{user_num1.class} or #{user_num2.class} type."
+    next
+
```
Comment (khjrtbrg) [inline calc.rb]: The segment of code that's executed in this conditional looks very similar to the code above it. You might consider combining it with the one above.

File: calc.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  elsif user_num1 == "0" 
+      num1 = 0
+  elsif user_num2 == "0"
+      num2 = 0
+  end
```
Comment (khjrtbrg) [inline calc.rb:19]: `num1` or `num2` should already be 0 at this point in your code, since you set them to `user_num1.to_f` and `user_num2.to_f` earlier.

File: calc.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    puts "You did not enter a valid operation."
+    next
+  end
+  #Changes answer to a float if necessary. 
+  if answer.to_f % 1 == 0
+    
+    num1 = num1.to_i
+    num2 = num2.to_i
+    answer = answer.to_i
+  end
```
Comment (khjrtbrg) [inline calc.rb:61]: Ah, I skimmed this section earlier and misread this. Are you were converting your answers to integers here for stylistic reasons?

File: calc.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+  elsif operation == "%" || operation == "mod" || operation == "modulo"
+    answer = num1 % num2
+    
+  else
+    puts "You did not enter a valid operation."
+    next
+  end
+  #Changes answer to a float if necessary. 
+  if answer.to_f % 1 == 0
```
Comment (khjrtbrg) [inline calc.rb]: It looks like your answer variable would almost always be converted to a float automatically at this point, so this conversion might not be necessary.

File: calc.rb [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  elsif operation == "**" || operation == "exponent" || operation == "^"
+    answer = num1 ** num2
+
+  elsif operation == "%" || operation == "mod" || operation == "modulo"
+    answer = num1 % num2
+    
+  else
+    puts "You did not enter a valid operation."
+    next
+  end
```
Comment (khjrtbrg) [inline calc.rb:53]: It might be nice to standardize your operation variable inside these conditions, so you don't end up with a result like "The result is 9 addition 1 = 10".

File: calc.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+  elsif operation == "**" || operation == "exponent" || operation == "^"
+    answer = num1 ** num2
+
+  elsif operation == "%" || operation == "mod" || operation == "modulo"
+    answer = num1 % num2
+    
+  else
+    puts "You did not enter a valid operation."
+    next
```
Comment (khjrtbrg) [inline calc.rb:52]: It feels a little jarring to be told my operator isn't valid after I've entered it and the numbers I want to evaluate. You might consider checking if the operator is valid before prompting for numbers, perhaps by checking against some sort of collection of options. It's convenient to siphon your calculations and check if it's a valid operator in one conditional, but that might be giving this if statement too much responsibility. Is it here to calculate the answer or is it here to check if the operator is valid? Keeping your code in small, intentional chunks can only help in the long run!

**Comment by khjrtbrg** [comment]
Looks good! I like how you collected your variables before outputting your final string, as well as how you skipped to the next loop if the user gave you a bad input. It was a little hard to follow all the conversions of variables to different datatypes, so that might be an area to clarify and refactor. Let me know if you have any questions, or if you want me to take another look! :palm_tree:

File: calc.rb (hunk 2 above)
Comment (ktrops) [inline calc.rb]: Oh I see. I honestly don't remember why I did that, maybe I didn't know about unless instead of using if to test if the answer needed to be a float? I'll change that.

File: calc.rb (hunk 3 above)
Comment (ktrops) [inline calc.rb]: I now see that's totally redundant. I could use just one variable keep it as a string until I'm done testing user input then once that passes change it to an integer.

File: calc.rb (hunk 4 above)
Comment (ktrops) [inline calc.rb]: I'll change this too. I think I was just uncomfortable with too many &&

File: calc.rb (hunk 6 above)
Comment (ktrops) [inline calc.rb:61]: Yes, but I could do :  
unless answer.to_f % 1 == 0  
    num1 = num1.to_f
    num2 = num2.to_f
    answer = answer.to_f
  end

Comment (khjrtbrg) [inline calc.rb:61]: That unless condition is a little hard for me to follow. I think if you're concerned with presentation, your previous approach of working with floats and converting to integers is easier to read and understand. That being said, I don't see an issue with returning your result in the data type Ruby outputs, unless one of your instructors asked you to do otherwise.

**Comment by ktrops** [comment]
I made some changes based on your comments. Thanks so much for your review!
